Support ==,!=,<,<=,>,>= over vectors

tautschnig opened the pull request:
These need to be rewritten to point-wise applications and return -1
(true) or 0 (false) per element.

Fixes: #5930

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening commented:
What happens when the vector subtype is unsigned?

tautschnig commented:
> What happens when the vector subtype is unsigned?

GCC's documentation says "constant of the appropriate type where all bits are set" (https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html), so this should work (as we use two's complement).

tautschnig commented on src/util/std_expr.h:
```diff
@@ -664,8 +664,8 @@ class binary_predicate_exprt:public binary_exprt
       vm,
-      expr.type().id() == ID_bool,
-      "result of binary predicate expression should be of type bool");
+      expr.type().id() == ID_bool || expr.type().id() == ID_vector,
+      "result of binary predicate expression should be of type bool or vector");
```
@kroening I quite dislike that I had to make the above change: this is really to make `--validate-goto-model` pass, which is currently applied before and (!) after calls to `remove_vector`. We should probably discuss this.

kroening replied:
Yes, not super. One could argue that something that returns a vector isn't a predicate.

tautschnig replied:
Perhaps the better solution would be introducing ids along the lines of `vector_equal`, `vector_gt`, etc?

kroening replied:
Not absolutely necessary -- we do overload many other IDs, but I admit that others may be surprised to see a non-boolean equality. `vector_equal` etc. may then be the best compromise.

kroening commented:
> > What happens when the vector subtype is unsigned?
> 
> GCC's documentation says "constant of the appropriate type where all bits are set" (https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html), so this should work (as we use two's complement).

Does that operation exist for vectors of type `float`?

tautschnig commented:
> > > What happens when the vector subtype is unsigned?
> > 
> > 
> > GCC's documentation says "constant of the appropriate type where all bits are set" (https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html), so this should work (as we use two's complement).
> 
> Does that operation exist for vectors of type `float`?

The docs say: "Comparison operands can be vector expressions of integer-type or real-type."

kroening commented:
> > Does that operation exist for vectors of type `float`?
> 
> The docs say: "Comparison operands can be vector expressions of integer-type or real-type."

Won't the `from_integer(-1, subtype)` then fail?

kroening commented:
> > The docs say: "Comparison operands can be vector expressions of integer-type or real-type."
> 
> Won't the `from_integer(-1, subtype)` then fail?

`from_integer(-1, subtype)` actually works for `float`. However, you then do get the floating-point representation of -1, which is not the same as "constant of the appropriate type where all bits are set" -- that would be NaN.

kroening commented:
> `from_integer(-1, subtype)` actually works for `float`. However, you then do get the floating-point representation of -1, which is not the same as "constant of the appropriate type where all bits are set" -- that would be NaN.

A quick test confirms that clang generates NaN, not -1.

tautschnig commented:
> > `from_integer(-1, subtype)` actually works for `float`. However, you then do get the floating-point representation of -1, which is not the same as "constant of the appropriate type where all bits are set" -- that would be NaN.
> 
> A quick test confirms that clang generates NaN, not -1.

Ah, just noting that `subtype` (which is the type of the result, not that of the operands) wouldn't be `float` here, it would still be `signed_int`. Though that's only what the C front-end generates, I acknowledge that I omitted any checking in here.

kroening commented:
Ah, I overlooked "The result of the comparison is a vector of the same width and number of elements as the comparison operands with a signed integral element type".

Right now, the result type is an array type, but should be a vector type.

kroening commented:
Ah, sorry, wrong again -- the frontend does produce a vector type, of signed integers, but it does fail to meet the requirement "same width".

tautschnig committed:
remove_vector: maintain source location

Newly created expressions did not have a source location, which may,
however, be useful in debugging.

tautschnig committed:
Fix typechecking of relational expressions over vectors

As specified in GCC's documentation, the result type is a vector of
signed integers of the same width as the operand subtypes. As the
semantics are different from relational expressions over scalars,
introduce a different expression id: each operation is now prefixed with
vector-.

tautschnig committed:
Support ==,!=,<,<=,>,>= over vectors

These need to be rewritten to point-wise applications and return -1
(true) or 0 (false) per element.

Fixes: #5930

kroening commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1437,9 +1431,18 @@ void c_typecheck_baset::typecheck_expr_rel_vector(
+    expr.id(ID_vector_notequal);
+  else
+    expr.id("vector-" + id2string(expr.id()));
 }
```
This will be vector-=, vector-<= etc?
